pkg/mirroring/push.go

```diff
@@ -0,0 +1,109 @@
+package mirroring
+
+import (
+	"os"
+
+	"github.com/serum-errors/go-serum"
+	"github.com/warptools/warpforge/pkg/logging"
+	"github.com/warptools/warpforge/pkg/workspace"
+	"github.com/warptools/warpforge/wfapi"
+)
+
+type pusher interface {
+	// Errors:
+	//
+	// 	- warpforge-error-io -- for IO errors that occur during push operations
+	hasWare(wfapi.WareID) (bool, error)
+	// Errors:
+	//
+	// 	- warpforge-error-io -- for IO errors that occur during push operations
+	pushWare(wfapi.WareID, string) error
+}
+
+func pusherFromConfig(cfg wfapi.WarehouseMirroringConfig) (pusher, error) {
+	if cfg.PushConfig.S3 != nil {
+		pusher, err := newS3Pusher(*cfg.PushConfig.S3)
+		return &pusher, err
+	} else if cfg.PushConfig.Mock != nil {
+		pusher, err := newMockPusher(*cfg.PushConfig.Mock)
+		return &pusher, err
+	} else {
+		// this should be unreachable due to IPLD validation
+		panic("no supported push configuration provided")
+	}
+}
+
+// PushToWarehouseAddr puts files into a mirror
+//
+// It requires a workspace and catalog to operate on, and the address and configuration
+// for mirroring. The given catalog will be scanned for wares that can be pushed based on
+// the configuration.
+//
+// Errors:
+//
+// 	- warpforge-error-io -- for IO errors that occur during push operations
+//  - warpforge-error-catalog-invalid -- when the provided catalog contains invalid data
+//  - warpforge-error-catalog-missing-entry -- should never occur, as we iterate over the contents of the catalog
+//  - warpforge-error-catalog-parse -- when the provided catalog cannot be parsed
+func PushToWarehouseAddr(log *logging.Logger, ws workspace.Workspace, cat workspace.Catalog, pushAddr wfapi.WarehouseAddr, cfg wfapi.WarehouseMirroringConfig) error {
```

Comment:
The logger should be configured in CLI layer and passed down in `context.Context`.